harvester/harvester
# feat(backup): upload vmbackup metadata

**Problem:**
We didn't store VMBackup metadata, so we can't redeploy to other clusters.

**Solution:**
Save VMBackup metadata to the backup target.

**Related Issue:**
https://github.com/harvester/harvester/issues/988

**Test plan:**
Case 1 - Sync VMBackup Metadata back.
* Setup an NFS/S3 backup target.
* Create a VM `vm`.
* Modify a file like `tmp.txt` and input `123` in it.
* Create a VM Backup `backup1`.
* Modify a file like `tmp.txt` and change content to `456`.
* Create a VM Backup `backup2`.
  * Check associated VolumeSnapshot / VolumeSnapshotContent/ LH Backup exist.
  * Check `default-backup1.cfg` and `default-backup2.cfg` are in the backup target.
  * Check `LonghornBackupName` exists in VM Backup `backup1` and `backup2`.
  * Check `LonghornBackupName` in CR `backup1` and `backup2` is same as `backup1.cfg` and `backup2.cfg` in backup target.
* Change backup target to another NFS/S3 backup target.
* Delete the VM Backup `backup1` and `backup2` and VM `vm`.
  * Check associated VolumeSnapshot / VolumeSnapshotContent / LH Backup are deleted.
  * Check `default-backup1.cfg` and `default-backup2.cfg` are still in the original backup target.
* Change backup target back.
  * Check VMBackup `backup1` and `backup2` and associated VolumeSnapshot / VolumeSnapshotContent / LH Backup are created again.
* Restore `backup1` to `vm1`
  * Check whether the file has `123` in it.
* Restore `backup2` to `vm2`
  * Check whether the file has `456` in it.

Case 2 - Delete VM Backup should delete the remote file.
* Setup an NFS/S3 backup target.
* Create a VM and create a VM Backup `backup1`.
  * Check whether `default-backup1.cfg` is in the backup target.
* Delete the VM Backup.
  * Check whether `default-backup1.cfg` is deleted in the backup target.

Case 3 - Delete VirtualMachineImage and restore VM. **(Not support in this PR. Just leave the test case as a reminder)**
* Follow Case 1, but also delete `VirtualMachineImage` after change backup target.
* Create `VirtualMachineImage` again with same `BackingImage` name.
* Restore VM Backup to VM. VM should run successfully.

**Environment Setup**
We move backup target validation to webhook and update `VirtualMachineBackup` CRD. If users would like to test the PR before it merged, please remember to update `harvester-webhook` deployment and `VirtualMachineBackup` CRD.

![Screen Shot 2021-11-23 at 5 32 58 PM](https://user-images.githubusercontent.com/4927361/143000363-937a1d61-eec9-4b36-a919-e5d653aae4e9.png)

![Screen Shot 2021-11-23 at 5 33 19 PM](https://user-images.githubusercontent.com/4927361/143000413-c35bc0c6-29bb-462c-9a2f-c0f05a640f2f.png)
